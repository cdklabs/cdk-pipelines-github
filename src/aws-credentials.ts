import { awsCredentialStep } from './private/aws-credentials';
import * as github from './workflows-model';

/**
 * AWS credential provider
 */
export abstract class AwsCredentialsProvider {
  public abstract jobPermission(): github.JobPermission;
  public abstract credentialSteps(region: string, assumeRoleArn?: string): github.JobStep[];
}

/**
 * Locations of GitHub Secrets used to authenticate to AWS
 */
export interface GitHubSecretsProviderProps {
  /**
   * @default "AWS_ACCESS_KEY_ID"
   */
  readonly accessKeyId: string;

  /**
   * @default "AWS_SECRET_ACCESS_KEY"
   */
  readonly secretAccessKey: string;

  /**
   * @default - no session token is used
   */
  readonly sessionToken?: string;
}

/**
 * AWS credential provider from GitHub secrets
 */
class GitHubSecretsProvider extends AwsCredentialsProvider {
  private readonly accessKeyId: string;
  private readonly secretAccessKey: string;
  private readonly sessionToken?: string;

  constructor(props?: GitHubSecretsProviderProps) {
    super();
    this.accessKeyId = props?.accessKeyId ?? 'AWS_ACCESS_KEY_ID';
    this.secretAccessKey = props?.secretAccessKey ?? 'AWS_SECRET_ACCESS_KEY';
    this.sessionToken = props?.sessionToken;
  }

  public jobPermission(): github.JobPermission {
    return github.JobPermission.NONE;
  }

  public credentialSteps(region: string, assumeRoleArn?: string): github.JobStep[] {
    return [
      awsCredentialStep('Authenticate Via GitHub Secrets', {
        region,
        accessKeyId: `\${{ secrets.${this.accessKeyId} }}`,
        secretAccessKey: `\${{ secrets.${this.secretAccessKey} }}`,
        ...(this.sessionToken ? {
          sessionToken: `\${{ secrets.${this.sessionToken} }}`,
        } : undefined),
        ...(assumeRoleArn ? {
          roleToAssume: assumeRoleArn,
          roleExternalId: 'Pipeline',
        } : undefined),
      }),
    ];
  }
}

/**
 * Role to assume using OpenId Connect
 */
export interface OpenIdConnectProviderProps {
  /**
   * A role that utilizes the GitHub OIDC Identity Provider in your AWS account.
   *
   * You can create your own role in the console with the necessary trust policy
   * to allow gitHub actions from your gitHub repository to assume the role, or
   * you can utilize the `GitHubActionRole` construct to create a role for you.
   */
  readonly gitHubActionRoleArn: string;
}

/**
 * AWS credential provider from OpenId Connect
 */
class OpenIdConnectProvider extends AwsCredentialsProvider {
  private readonly gitHubActionRoleArn: string;

  constructor(props: OpenIdConnectProviderProps) {
    super();
    this.gitHubActionRoleArn = props.gitHubActionRoleArn;
  }

  public jobPermission(): github.JobPermission {
    return github.JobPermission.WRITE;
  }

  public credentialSteps(region: string, assumeRoleArn?: string): github.JobStep[] {
    function getDeployRole(arn: string) {
      return arn.replace('cfn-exec', 'deploy');
    }

    let steps: github.JobStep[] = [];

    steps.push(
      awsCredentialStep('Authenticate Via OIDC Role', {
        region,
        roleToAssume: this.gitHubActionRoleArn,
      }),
    );

    if (assumeRoleArn) {
      // Result of initial credentials with GitHub Action role are these environment variables
      steps.push(
        awsCredentialStep('Assume CDK Deploy Role', {
          region,
          accessKeyId: '${{ env.AWS_ACCESS_KEY_ID }}',
          secretAccessKey: '${{ env.AWS_SECRET_ACCESS_KEY }}',
          sessionToken: '${{ env.AWS_SESSION_TOKEN }}',
          roleToAssume: getDeployRole(assumeRoleArn),
          roleExternalId: 'Pipeline',
        }),
      );
    }

    return steps;
  }
}

/**
 * Dummy AWS credential provider
 */
class NoCredentialsProvider extends AwsCredentialsProvider {
  public jobPermission(): github.JobPermission {
    return github.JobPermission.NONE;
  }
  public credentialSteps(_region: string, _assumeRoleArn?: string): github.JobStep[] {
    return [];
  }
}

/**
 * Provides AWS credenitals to the pipeline jobs
 */
export class AwsCredentials {
  /**
   * Reference credential secrets to authenticate with AWS. This method assumes
   * that your credentials will be stored as long-lived GitHub Secrets.
   */
  static fromGitHubSecrets(props?: GitHubSecretsProviderProps): AwsCredentialsProvider {
    return new GitHubSecretsProvider(props);
  }

  /**
   * Provide AWS credentials using OpenID Connect.
   */
  static fromOpenIdConnect(props: OpenIdConnectProviderProps): AwsCredentialsProvider {
    return new OpenIdConnectProvider(props);
  }

  /**
   * Don't provide any AWS credentials, use this if runners have preconfigured credentials.
   */
  static runnerHasPreconfiguredCreds(): AwsCredentialsProvider {
    return new NoCredentialsProvider();
  }
}
