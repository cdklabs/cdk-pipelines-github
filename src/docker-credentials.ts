/**
 * Represents a credential used to authenticate to a docker registry.
 * Uses the official Docker Login GitHub Action to authenticate.
 *
 * @see https://github.com/marketplace/actions/docker-login
 */
export class DockerCredential {
  /**
   * Reference credential secrets to authenticate to DockerHub. This method assumes
   * that your credentials will be stored as long-lived GitHub Secrets under the
   * usernameKey and personalAccessTokenKey.
   *
   * The default for usernameKey is `DOCKERHUB_USERNAME`. The default for personalAccessTokenKey
   * is `DOCKERHUB_TOKEN`. If you do not set these values, your credentials should be
   * found in your GitHub Secrets under these default keys.
   */
  public static dockerHub(creds: DockerHubCredentialSecrets = {}): DockerCredential {
    return new DockerCredential(
      'docker',
      undefined,
      creds.usernameKey ?? 'DOCKERHUB_USERNAME',
      creds.personalAccessTokenKey ?? 'DOCKERHUB_TOKEN',
    );
  }

  /**
   * Create a credential for ECR. This method will reuse your AWS credentials to log in to AWS.
   * Your AWS credentials are already used to deploy your CDK stacks. It can be supplied via
   * GitHub Secrets or using an IAM role that trusts the GitHub OIDC identity provider.
   *
   * NOTE - All ECR repositories in the same account and region share a domain name
   * (e.g., 0123456789012.dkr.ecr.eu-west-1.amazonaws.com), and can only have one associated
   * set of credentials (and DockerCredential). Attempting to associate one set of credentials
   * with one ECR repo and another with another ECR repo in the same account and region will
   * result in failures when using these credentials in the pipeline.
   */
  public static ecr(registry: string): DockerCredential {
    return new DockerCredential('ecr', registry);
  }

  /**
   * Create a credential for a custom registry. This method assumes that you will have long-lived
   * GitHub Secrets stored under the usernameKey and passwordKey that will authenticate to the
   * registry you provide.
   *
   * @see https://github.com/marketplace/actions/docker-login
   */
  public static customRegistry(registry: string, creds: ExternalDockerCredentialSecrets): DockerCredential {
    return new DockerCredential('custom', registry, creds.usernameKey, creds.passwordKey);
  }

  private constructor(
    readonly name: string,
    readonly registry?: string,
    readonly usernameKey?: string,
    readonly passwordKey?: string,
  ) {}
}

/**
 * Locations of GitHub Secrets used to authenticate to DockerHub.
 */
export interface DockerHubCredentialSecrets {
  /**
   * The key of the GitHub Secret containing the DockerHub username.
   *
   * @default 'DOCKERHUB_USERNAME'
   */
  readonly usernameKey?: string;

  /**
   * The key of the GitHub Secret containing the DockerHub personal access token.
   *
   * @default 'DOCKERHUB_TOKEN'
   */
  readonly personalAccessTokenKey?: string;
}

/**
 * Generic structure to supply the locations of GitHub Secrets used to authenticate
 * to a docker registry.
 */
export interface ExternalDockerCredentialSecrets {
  /**
   * The key of the GitHub Secret containing your registry username.
   */
  readonly usernameKey: string;

  /**
   * The key of the GitHub Secret containing your registry password.
   */
  readonly passwordKey: string;
}
