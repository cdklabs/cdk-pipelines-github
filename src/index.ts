export * from './pipeline';
export * from './workflows-model';
export * from './oidc-provider';
export * from './docker-credentials';
export * from './stage-options';
