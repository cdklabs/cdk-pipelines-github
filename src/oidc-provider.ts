import { Aws, CfnOutput } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

/**
 * GitHub OIDC thumbprints updated 2023-07-27
 *
 * https://github.blog/changelog/2023-06-27-github-actions-update-on-oidc-integration-with-aws/
 */
const GITHUB_OIDC_THUMBPRINTS = [
  '6938fd4d98bab03faadb97b34396831e3780aea1',
  '1c58a3a8518e8759bf075b76b750d4f2df264fcd',
];

/**
 * Properties for the GitHubActionRole construct.
 */
export interface GitHubActionRoleProps {
  /**
   * A list of GitHub repositories you want to be able to access the IAM role.
   * Each entry should be your GitHub username and repository passed in as a
   * single string.
   * An entry `owner/repo` is equivalent to the subjectClaim `repo:owner/repo:*`.
   *
   * For example, `['owner/repo1', 'owner/repo2'].
   */
  readonly repos?: string[];

  /**
   * A list of subject claims allowed to access the IAM role.
   * See https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect
   * A subject claim can include `*` and `?` wildcards according to the `StringLike`
   * condition operator.
   *
   * For example, `['repo:owner/repo1:ref:refs/heads/branch1', 'repo:owner/repo1:environment:prod']`
   */
  readonly subjectClaims?: string[];

  /**
   * The name of the Oidc role.
   *
   * @default 'GitHubActionRole'
   */
  readonly roleName?: string;

  /**
   * The GitHub OpenId Connect Provider. Must have provider url
   * `https://token.actions.githubusercontent.com`. The audience must be
   * `sts:amazonaws.com`.
   *
   * Only one such provider can be defined per account, so if you already
   * have a provider with the same url, a new provider cannot be created for you.
   *
   * @default - a provider is created for you.
   */
  readonly provider?: iam.IOpenIdConnectProvider;

  /**
   * Thumbprints of GitHub's certificates
   *
   * Every time GitHub rotates their certificates, this value will need to be updated.
   *
   * Default value is up-to-date to June 27, 2023 as per
   * https://github.blog/changelog/2023-06-27-github-actions-update-on-oidc-integration-with-aws/
   *
   * @default - Use built-in keys
   */
  readonly thumbprints?: string[];
}

/**
 * Creates or references a GitHub OIDC provider and accompanying role that trusts the provider.
 * This role can be used to authenticate against AWS instead of using long-lived AWS user credentials
 * stored in GitHub secrets.
 *
 * You can do this manually in the console, or create a separate stack that uses this construct.
 * You must `cdk deploy` once (with your normal AWS credentials) to have this role created for you.
 *
 * You can then make note of the role arn in the stack output and send it into the Github Workflow app via
 * the `gitHubActionRoleArn` property. The role arn will be `arn:<partition>:iam::<accountId>:role/GithubActionRole`.
 *
 * @see https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services
 */
export class GitHubActionRole extends Construct {
  /**
   * Reference an existing GitHub Actions provider.
   * You do not need to pass in an arn because the arn for such
   * a provider is always the same.
   */
  public static existingGitHubActionsProvider(scope: Construct): iam.IOpenIdConnectProvider {
    return iam.OpenIdConnectProvider.fromOpenIdConnectProviderArn(
      scope,
      'GitHubActionProvider',
      `arn:${Aws.PARTITION}:iam::${Aws.ACCOUNT_ID}:oidc-provider/token.actions.githubusercontent.com`,
    );
  }

  /**
   * The role that gets created.
   *
   * You should use the arn of this role as input to the `gitHubActionRoleArn`
   * property in your GitHub Workflow app.
   */
  public readonly role: iam.IRole;

  constructor(scope: Construct, id: string, props: GitHubActionRoleProps) {
    super(scope, id);

    const rawEndpoint = 'token.actions.githubusercontent.com';
    const providerUrl = `https://${rawEndpoint}`;

    // uses the given provider or creates a new one.
    const provider = props.provider ?? new iam.OpenIdConnectProvider(this, 'github-provider', {
      url: providerUrl,
      clientIds: ['sts.amazonaws.com'],
      thumbprints: props.thumbprints ?? GITHUB_OIDC_THUMBPRINTS,
    });

    // create a role that references the provider as a trusted entity
    const principal = new iam.FederatedPrincipal(
      provider.openIdConnectProviderArn,
      {
        StringLike: {
          [`${rawEndpoint}:sub`]: formatRepos(props.repos ?? []).concat(props.subjectClaims ?? []),
        },
      },
      'sts:AssumeRoleWithWebIdentity',
    );

    // permit this role from assuming all of the CDK bootstrap roles
    const oidcPolicyStatement = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['sts:AssumeRole'],
      resources: ['*'],
      conditions: {
        'ForAnyValue:StringEquals': {
          'iam:ResourceTag/aws-cdk:bootstrap-role': [
            'deploy',
            'lookup',
            'file-publishing',
            'image-publishing',
          ],
        },
      },
    });

    // permit this role from accessing ecr repositories for docker assets
    const ecrPolicyStatement = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['ecr:GetAuthorizationToken'],
      resources: ['*'],
    });

    this.role = new iam.Role(this, 'github-action-role', {
      roleName: props.roleName ?? 'GitHubActionRole',
      assumedBy: principal,
      inlinePolicies: {
        AssumeBootstrapRoles: new iam.PolicyDocument({
          statements: [oidcPolicyStatement, ecrPolicyStatement],
        }),
      },
    });

    // show the role arn in the stack output
    new CfnOutput(this, 'roleArn', {
      value: this.role.roleArn,
    });
  }
}

function formatRepos(repos: string[]) {
  const formattedRepos = [];
  for (const repo of repos) {
    formattedRepos.push(`repo:${repo}:*`);
  }
  return formattedRepos;
}
