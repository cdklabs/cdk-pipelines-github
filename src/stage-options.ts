import { AddStageOpts } from 'aws-cdk-lib/pipelines';

/**
 * Options to pass to `addStageWithGitHubOpts`.
 */
export interface AddGitHubStageOptions extends AddStageOpts {
  /**
   * Run the stage in a specific GitHub Environment. If specified,
   * any protection rules configured for the environment must pass
   * before the job is set to a runner. For example, if the environment
   * has a manual approval rule configured, then the workflow will
   * wait for the approval before sending the job to the runner.
   *
   * Running a workflow that references an environment that does not
   * exist will create an environment with the referenced name.
   * @see https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment
   *
   * @default - no GitHub environment
   */
  readonly gitHubEnvironment?: string;

  /**
   * In some cases, you must explicitly acknowledge that your CloudFormation
   * stack template contains certain capabilities in order for CloudFormation
   * to create the stack.
   *
   * If insufficiently specified, CloudFormation returns an `InsufficientCapabilities`
   * error.
   *
   * @default ['CAPABILITY_IAM']
   */
  readonly stackCapabilities?: StackCapabilities[];
}

/**
 * Acknowledge IAM resources in AWS CloudFormation templates.
 *
 * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-iam-template.html#capabilities
 */
export enum StackCapabilities {
  /** Acknowledge your stack includes IAM resources */
  IAM = 'CAPABILITY_IAM',

  /** Acknowledge your stack includes custom names for IAM resources */
  NAMED_IAM = 'CAPABILITY_NAMED_IAM',

  /** Acknowledge your stack contains one or more macros */
  AUTO_EXPAND = 'CAPABILITY_AUTO_EXPAND',
}