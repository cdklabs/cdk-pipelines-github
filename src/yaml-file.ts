import { writeFileSync } from 'fs';
import * as YAML from 'yaml';
import { JsonPatch } from './json-patch';

/**
 * Options for `YamlFile`
 */
export interface YamlFileOptions {
  /**
   * The object that will be serialized. You can modify the object's contents
   * before synthesis.
   *
   * @default {} an empty object
   */
  readonly obj?: any;
}

/**
 * Represents a Yaml File.
 */
export class YamlFile {
  /**
   * The path to the file that the object will be written to.
   */
  private readonly filePath: string;

  /**
   * The output object. This object can be mutated until the project is synthesized.
   */
  private obj: object;

  /**
   * Patches to be applied to `obj` after the resolver is called.
   */
  private readonly patchOperations: JsonPatch[];

  /**
   * A comment to be added to the top of the YAML file.
   *
   * Can be multiline. All non-empty line are pefixed with '# '. Empty lines are kept, but not commented.
   *
   * For example:
   * ```ts
   * declare const pipeline: GitHubWorkflow;
   * pipeline.workflowFile.commentAtTop =
   * `AUTOGENERATED FILE, DO NOT EDIT!
   * See ReadMe.md
   * `;
   * ```
   *
   * Results in YAML:
   * ```yaml
   * # AUTOGENERATED FILE, DO NOT EDIT!
   * # See ReadMe.md
   *
   * name: deploy
   * ...
   * ```
   */
  public commentAtTop?: string;

  constructor(filePath: string, options: YamlFileOptions = {}) {
    this.filePath = filePath;
    this.obj = options.obj ?? {};
    this.patchOperations = [];

    // default value for commentAtTop
    this.commentAtTop = `AUTOMATICALLY GENERATED FILE, DO NOT EDIT MANUALLY.
Generated by AWS CDK and [cdk-pipelines-github](https://github.com/cdklabs/cdk-pipelines-github)`;
  }

  /**
   * Update the output object.
   */
  public update(obj: any) {
    this.obj = obj;
  }

  /**
   * Applies an RFC 6902 JSON-patch to the synthesized object file.
   * See https://datatracker.ietf.org/doc/html/rfc6902 for more information.
   *
   * For example, with the following yaml file
   * ```yaml
   * name: deploy
   * on:
   *   push:
   *     branches:
   *       - main
   *   workflow_dispatch: {}
   * ...
   * ```
   *
   * modified in the following way:
   *
   * ```ts
   * declare const pipeline: GitHubWorkflow;
   * pipeline.workflowFile.patch(JsonPatch.add("/on/workflow_call", "{}"));
   * pipeline.workflowFile.patch(JsonPatch.remove("/on/workflow_dispatch"));
   * ```
   *
   * would result in the following yaml file:
   *
   * ```yaml
   * name: deploy
   * on:
   *   push:
   *     branches:
   *       - main
   *   workflow_call: {}
   * ...
   * ```
   *
   * @param patches - The patch operations to apply
   */
  public patch(...patches: JsonPatch[]) {
    this.patchOperations.push(...patches);
  }

  /**
   * Returns the patched yaml file.
   */
  public toYaml(): string {
    const obj = JSON.parse(JSON.stringify(this.obj));
    const patched = JsonPatch.apply(obj, ...this.patchOperations);

    const yamlDoc = new YAML.Document(patched);
    yamlDoc.commentBefore = this.commentAtTop ?? null;

    return yamlDoc.toString({
      commentString: (comment) =>
        comment.split('\n').map((x) => x == '' ? '' : `# ${x}`).join('\n'),
      indent: 2,
    });
  }

  /**
   * Write the patched yaml file to the specified location.
   */
  public writeFile() {
    writeFileSync(this.filePath, this.toYaml());
  }
}
