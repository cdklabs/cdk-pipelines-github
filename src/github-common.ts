import { AddStageOpts } from 'aws-cdk-lib/pipelines';
import { JobSettings } from './pipeline';
import { StackCapabilities } from './stage-options';

/**
 * Common properties to extend both StageProps and AddStageOpts
 */

export interface GitHubCommonProps {
  /**
   * Run the stage in a specific GitHub Environment. If specified,
   * any protection rules configured for the environment must pass
   * before the job is set to a runner. For example, if the environment
   * has a manual approval rule configured, then the workflow will
   * wait for the approval before sending the job to the runner.
   *
   * Running a workflow that references an environment that does not
   * exist will create an environment with the referenced name.
   * @see https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment
   *
   * @default - no GitHub environment
   */
  readonly gitHubEnvironment?: string;

  /**
   * In some cases, you must explicitly acknowledge that your CloudFormation
   * stack template contains certain capabilities in order for CloudFormation
   * to create the stack.
   *
   * If insufficiently specified, CloudFormation returns an `InsufficientCapabilities`
   * error.
   *
   * @default ['CAPABILITY_IAM']
   */
  readonly stackCapabilities?: StackCapabilities[];

  /**
   * Job level settings that will be applied to all jobs in the stage.
   * Currently the only valid setting is 'if'.
   */
  readonly jobSettings?: JobSettings;
}

/**
 * Options to pass to `addStageWithGitHubOpts`.
 */
export interface AddGitHubStageOptions extends AddStageOpts, GitHubCommonProps {}
