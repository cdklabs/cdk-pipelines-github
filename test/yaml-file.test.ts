import { readFileSync } from 'fs';
import * as path from 'path';
import * as YAML from 'yaml';
import { JsonPatch } from '../src/json-patch';
import { YamlFile } from '../src/yaml-file';
import { withTemporaryDirectory } from './testutil';

describe('patch', () => {
  function patchTest(patches: JsonPatch[], initialObj: any, assertObj: any, updateObj?: any) {
    withTemporaryDirectory((dir) => {
      const fileName = path.join(dir, 'file.yml');
      const yamlFile = new YamlFile(fileName, {
        obj: initialObj,
      });
      yamlFile.patch(...patches);

      if (updateObj) {
        yamlFile.update(updateObj);
      }

      yamlFile.writeFile();

      expect(YAML.parse(readFileSync(fileName, 'utf-8'))).toStrictEqual(assertObj);
    });
  }

  test('patch(p, v) can add to an existing array', () => {
    patchTest(
      [
        JsonPatch.add('/first/second/array/-', '1'),
        JsonPatch.add('/first/second/array/-', '2'),
        JsonPatch.add('/first/second/array/1', '3'),
      ],
      { first: { second: { array: ['0'] } } },
      { first: { second: { array: ['0', '3', '1', '2'] } } },
    );
  });

  test('patch(p, v) can add to an existing array after update', () => {
    patchTest(
      [
        JsonPatch.add('/first/second/array/-', '1'),
        JsonPatch.add('/first/second/array/-', '2'),
        JsonPatch.add('/first/second/array/1', '3'),
      ],
      { first: { second: { array: ['?'] } } }, // shouldn't use
      { first: { second: { array: ['0', '3', '1', '2'] } } },
      { first: { second: { array: ['0'] } } }, // should use
    );
  });

  test('patch(p, v) can create an array', () => {
    patchTest(
      [
        JsonPatch.add('/first/second/array', []),
        JsonPatch.add('/first/second/array/-', '1'),
        JsonPatch.add('/first/second/array/-', '2'),
      ],
      { first: { third: {} } }, // shouldn't use
      { first: { second: { array: ['1', '2'] } } },
      { first: { second: {} } }, // should use
    );
  });
});


describe('yaml file comments', () => {
  function commentTest(commentAtTop: string | undefined, initialObj: any, assetDoc: string) {
    withTemporaryDirectory((dir) => {
      const fileName = path.join(dir, 'file.yml');
      const yamlFile = new YamlFile(fileName, {
        obj: initialObj,
      });
      yamlFile.commentAtTop = commentAtTop;
      yamlFile.writeFile();

      expect(readFileSync(fileName, 'utf-8')).toMatchSnapshot(assetDoc);
      expect(yamlFile.commentAtTop).toStrictEqual(commentAtTop);
    });
  }

  test('comment at top works', () => {
    const commentBeforeYaml = 'commentAtTopTest.yml';
    commentTest(
      'Comment before',
      { first: { second: { array: ['0'] } } },
      commentBeforeYaml,
    );
  });

  test('multi-line comments work', () => {
    const commentAfterYaml = 'multilineCommentAtTopAndBottomTest.yml';
    commentTest(
      `Comment before:\n\n${['A', 'B', 'C'].map((x) => ` - ${x}\n`).join('')}`,
      { first: { second: { array: ['0'] } } },
      commentAfterYaml,
    );
  });
});