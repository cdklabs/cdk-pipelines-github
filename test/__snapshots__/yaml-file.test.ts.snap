// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yaml file comments comment at top works: commentAtTopTest.yml 1`] = `
"# Comment before

first:
  second:
    array:
      - \\"0\\"
"
`;

exports[`yaml file comments multi-line comments work: multilineCommentAtTopAndBottomTest.yml 1`] = `
"# Comment before:

#  - A
#  - B
#  - C


first:
  second:
    array:
      - \\"0\\"
"
`;
